sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,974 @@
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the
+sycl_ext_oneapi_free_function_kernels extension provides that mechanism.
+The "free function kernel" feature is a separate extension, though, because it
+can also be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_FREE_FUNCTION_KERNELS`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a free function kernel
+
+A free function kernel is a normal C++ function definition, where the function
+declaration is decorated with one of the following compile-time properties:
+`range_kernel`, `nd_range_kernel`, or `single_task_kernel`.
+
+When a function declaration is decorated with one of these properties, the
+following rules must be observed:
+
+* The function must be declared at either namespace scope or at class scope as
+  a static member function.
+
+* The function's return type must be `void`.
+
+* The function must not accept variadic arguments.
+
+* Each of the function's arguments must have a type that is an allowed kernel
+  parameter type as specified in section 4.12.4 "Rules for parameter passing to
+  kernels" of the core SYCL specification.
+  The function must not be declared with parameters of type `reducer` or
+  `kernel_handler`.
+  These special kernel arguments cannot be passed to a free function kernel.
+
+* All declarations of the function must provide no default parameter values.
+
+* The property must appear on the first declaration of the function in the
```

Comment:
Since these properties behave much like C++ attributes I would lean towards saying there should be a compile-time error in this case, as C++ compilers typically produce an error for the equivalent misuse of attributes. Whether we require this in the specification I think could be debated, though I would probably say it should, as the alternative would be much harder for users to diagnose.

Reply:
For now, I've captured this under "Implementation notes".  I agree that DPC++ should diagnose this error.  I also added an open issue about requiring these diagnostics for any implementation of this extension.  See 4d4535332b85e2e9ff88828b5f8baf9bd16a4a0d and 6ed0ec5e4f889c7c6b92f616313437e293a18ab9.